tool-openssl/verify.cc

```diff
@@ -0,0 +1,198 @@
+        /* Continue even if the leaf is a self-signed cert */
+      case X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT:
+        /* Continue after extension errors too */
+      case X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY:
+      case X509_V_ERR_INVALID_CA:
+      case X509_V_ERR_INVALID_NON_CA:
+      case X509_V_ERR_PATH_LENGTH_EXCEEDED:
+      case X509_V_ERR_CRL_HAS_EXPIRED:
+      case X509_V_ERR_CRL_NOT_YET_VALID:
+      case X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION:
+        /* errors due to strict conformance checking (-x509_strict) */
+      case X509_V_ERR_INVALID_PURPOSE:
+        ok = 1;
+    }
+  }
+  return ok;
+}
+
+static int check(X509_STORE *ctx, const char *file) {
+  bssl::UniquePtr<X509> x;
+  int i = 0, ret = 0;
+  X509_STORE_CTX *store_ctx;
+
+  if (file) {
+    ScopedFILE cert_file(fopen(file, "rb"));
+    if (!cert_file) {
+      fprintf(stderr, "error %s: reading certificate failed\n", file);
+    }
+    x.reset(PEM_read_X509(cert_file.get(), nullptr, nullptr, nullptr));
+
+  } else {
+    bssl::UniquePtr<BIO> input(BIO_new_fp(stdin, BIO_CLOSE));
+    x.reset(PEM_read_bio_X509(input.get(), NULL, NULL, NULL));
+  }
+
+  if (x.get() == NULL) {
+    return 0;
+  }
+
+  store_ctx = X509_STORE_CTX_new();
+  if (store_ctx == NULL) {
+    fprintf(stderr, "error %s: X.509 store context allocation failed\n",
+               (file == NULL) ? "stdin" : file);
+    return 0;
+  }
+
+  if (!X509_STORE_CTX_init(store_ctx, ctx, x.get(), NULL)) {
+    X509_STORE_CTX_free(store_ctx);
+    fprintf(stderr,
+               "error %s: X.509 store context initialization failed\n",
+               (file == NULL) ? "stdin" : file);
+    return 0;
+  }
+
+  i = X509_verify_cert(store_ctx);
+  if (i > 0 && X509_STORE_CTX_get_error(store_ctx) == X509_V_OK) {
+    fprintf(stdout, "%s: OK\n", (file == NULL) ? "stdin" : file);
+    ret = 1;
+  } else {
+    fprintf(stderr,
+               "error %s: verification failed\n",
+               (file == NULL) ? "stdin" : file);
+  }
+  X509_STORE_CTX_free(store_ctx);
+
+  return ret;
+}
+
+bool VerifyTool(const args_list_t &args) {
+  std::string cafile;
+  bssl::UniquePtr<X509_STORE> store;
+  int ret;
+  size_t i = 0;
+
+  if (args.size() == 1 && args[0] == "-help") {
+    fprintf(stderr, "Usage: verify [options] cert.pem...\n"
+                    "Certificates must be in PEM format and specified in a file.\n"
+                    "We do not support reading from stdin yet. \n\n "
+                    "Valid options are:\n");
+    PrintUsage(kArguments);
+    return false;
+  } else if (args.size() > 1 && args[0] == "-CAfile") {
+    cafile = args[1];
+    i += 2;
+  }
+
+  store.reset(setup_verification_store(cafile));
+
+  // Initialize certificate verification store
+  if (!store.get()) {
+    fprintf(stderr, "Error: Unable to setup certificate verification store.");
+    return false;
+  }
+  X509_STORE_set_verify_cb(store.get(), cb);
+
+  ERR_clear_error();
+
+  ret = 1;
```

Comment:
It would be better to assume failure (`ret = 0`), then only flip it to 1 on success.

Reply:
In the below code in the else block (lines 186-191):

The current logic, initializing `ret` to 1, is preferable. It allows us to:

1. Check all certificates without premature termination.
2. Efficiently track failures by setting `ret` to 0 on any single failure.
3. Avoid the complexity of tracking individual check results for multiple certs/files.

flipping the logic would require us to track every result and ensure they were all successful.

Reply:
I'm not sure why `ret` is needed at all.  The logic could just `return 0` in the places it currently has `ret = 0`. Is there a reason to continue checking after one of them fails?

Reply:
NP: A slightly less verbose (and clearer?) way to write this would be the following:

```
  ret = 1;

  if (args.size() == i) {
    ret &= check(store.get(), NULL);
  } else {
    // Certs provided as files
    for (; i < args.size(); i++) {
      ret &= check(store.get(), args[i].c_str());
    }
  }

  return ret == 1;
```

Reply:
agreed, made the change thank you!